### ISSUE TITLE
mob start with ci skip option

### ISSUE THREAD

**Issue by takaiyuk** [remotemobprogramming/mob#353]
## Context

#329  made `--push-option=ci.skip` available, which allows CI to be skipped in some use cases (perhaps only GitLab).
However, CI is not skipped when `mob start` in my organization, using CircleCI in GitHub.
This is probably because `--push-option=ci.skip` makes no sense except for GitLab.

## Proposal

- Introduce an opt-in option for CI skipping when `mob start`.
  - No change to default behavior due to opt-in option.
- Execute an empty commit with ci-skip message (e.g. `mob start [ci-skip] [ci skip] [skip ci]`) before push for skipping CI.
  - Commits are squashed when `mob done`, so there is no problem with empty commits.

### Implementation Example

- Execution command

```shell
mob start --ci-skip
```

or 

```shell
export MOB_START_WITH_CI_SKIP=true && mob start
```

- Output of command execution

```text
> starting new session from origin/BRANCH-NAME
  git checkout -B mob/BRANCH-NAME origin/BRANCH-NAME
  git commit --allow-empty -m mob start [ci-skip] [ci skip] [skip ci]
  git push --push-option ci.skip --no-verify --set-upstream origin mob/BRANCH-NAME
```

---

## Edit

- Make CI skip option be true as default.
- Initial empty commit with ci-skip message is removed when `mob done --squash-wip`.
  - not removed when `mob done --no-squash`
- `--push-option=ci.skip` to be deleted.

**Comment by gregorriegler** [comment]
Hi @takaiyuk,
Thank you a lot for your contribution!

So if I understand you correctly, the `git push --push-option ci.skip` does not help with skipping the CI in your environment.
However, if you did a `git commit --allow-empty -m mob start [ci-skip] [ci skip] [skip ci]` before, it would?
Is that so? Can you verify that for me?

In that case, I would propose that we skip the configuration property all together, and make the empty commit the default behaviour. As we always would want to skip the CI anyways. There would be no value in turning it off.

**Comment by takaiyuk** [comment]
Hi @gregorriegler, thank you for your quick response.

> So if I understand you correctly, the git push --push-option ci.skip does not help with skipping the CI in your environment.
However, if you did a git commit --allow-empty -m mob start [ci-skip] [ci skip] [skip ci] before, it would?
Is that so? Can you verify that for me?

Yes, you understand correctly.
It is hard to show you the verification but it helped us to skip CI because it’s similar to `mob next` commit message (`mob next [ci-skip] [ci skip] [skip ci]`), which skips CI for sure as you know.
(However, I will try to take a screenshot of the situation and share it with you.)

> In that case, I would propose that we skip the configuration property all together, and make the empty commit the default behaviour. As we always would want to skip the CI anyways. There would be no value in turning it off.

As you pointed out, I also think that it’s better to skip CI always than leaving the default behaviour as before.

> Another question that would be interesting is:
If the empty commit would solve the problem of skipping the CI, would there still be a reason to keep the --push-option ci.skip, or would this become obsolete?

I do not think `--push-option` is needed anymore. At least at the moment, whether this option properly works probably depends on the environment.

**Comment by takaiyuk** [comment]
> However, I will try to take a screenshot of the situation and share it with you.

If my latest commit is `Commit as usual` and execute`mob start`, then CI is triggered on the new mob branch as expected.

<img width="360" alt="screenshot1" src="https://user-images.githubusercontent.com/32270184/211981736-2406a063-f693-4f9a-89cf-bfb99ebb7d1d.png">

If you do a `git commit --allow-empty -m mob start [ci-skip] [ci skip] [skip ci]` before push, then CI is not triggered as when you run `mob next`.

<img width="417" alt="screenshot2" src="https://user-images.githubusercontent.com/32270184/211981772-043df58c-a07b-46c5-934a-cd2cfa23e3f4.png">

**Comment by simonharrer** [comment]
We should be clear what the consequences of such an empty WIP commit are.

Making this default makes using the `mob done --merge` option basically unusable.

**Comment by takaiyuk** [comment]
I cannot find out `--merge` option for `mob done`, but it is worth for us to check out what happens when `mob done` including the empty commit.

- `mob done` squashes commit messages, so there is no problem.
- `mob done --no-squash` leaves the empty commit.
- `mob done --squash` squashes commit messages, so there is no problem.
- `mob done --squash-wip` leaves the empty commit.

Some `mob done` options make the empty commit alive, so we need to consider them.

**Comment by gregorriegler** [comment]
Hi @takaiyuk, I think @simonharrer is talking about `mob done --squash-wip`.

The reason why mob `done --squash-wip` leaves the empty commit is that it is not programmed to replace a commit message named `mob start [ci-skip] [ci skip] [skip ci]`.
 It would however squash it if it was `mob next [ci-skip] [ci skip] [skip ci]` (subtle difference, **start** vs **next**)

We first need a failing test that uses `--squash-wip` and fails because the empty commit is not squashed whereas it expects all wip commits to be squashed including the empty one. 
Then we need to adapt `configuration.go`.
This function `IsWipCommitMessage` should also return true if the commit was named `mob start [ci-skip] [ci skip] [skip ci]`
https://github.com/remotemobprogramming/mob/blob/480227b04fa33684f849bbb2e85b85053b9693ca/configuration/configuration.go#L64

This should make the failing Test pass and solve our issue.

**Comment by takaiyuk** [comment]
Hi @gregorriegler, thanks for the comment in detail.

As you explained, `mob start [ci-skip] [ci skip] [skip ci]` should be squashed, but it is the first commit for the wip branch (since it is committed when `mob start`) and you cannot squash due to without a previous commit. So, I think that the commit have to be dropped instead of squashed.

**Comment by takaiyuk** [comment]
I got it with your step-by-step explanation. I will call `markStartCommitForDropping()` in `squashWipGitSequenceEditor()` for the return value of `markPostWipCommitsForSquashing()` and write test with `TestMarkSquashWip_manyWipCommitsFollowedByManualCommit()`.

OK, I will not drop the initial commit when `--no-squash` option.

**Comment by takaiyuk** [comment]
PR is ready: #354

**Comment by takaiyuk** [comment]
PR was merged and released as v4.2.0.
